### PR TITLE
feat(pageserver): support auto split layers based on size

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

part of https://github.com/neondatabase/neon/issues/8002

## Summary of changes

Add a `SplitImageWriter` that automatically splits image layer based on estimated target image layer size. This does not consider compression and we might need a better metrics.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 1]
```diff
@@ -980,6 +992,17 @@ impl ImageLayerWriter {
         self.inner.as_mut().unwrap().put_image(key, img, ctx).await
     }
 
+    #[cfg(test)]
+    /// Estimated size of the image layer. Currently, this is the blob (compressed) size without the index.
+    pub(crate) fn estimated_size(&self) -> u64 {
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/image_layer.rs:997]: in my unmerged PR, I've added a slightly more sophisticated function that adds the size of the index as well as the summary size: https://github.com/neondatabase/neon/pull/7879/files#diff-bb89836c560ca839968572f69cfbfaf2c7f72429e9c909e9c6329010b96ece53R1102 . maybe use that one?

Comment (skyzh) [reply pageserver/src/tenant/storage_layer/image_layer.rs:997]: sounds good, I'll change the estimation function

Comment (skyzh) [reply pageserver/src/tenant/storage_layer/image_layer.rs:997]: https://github.com/neondatabase/neon/pull/8574/commits/05b6c558b7f526897c44a7b744b8f2d9d2bacf73

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 2]
```diff
@@ -899,11 +905,17 @@ impl ImageLayerWriterInner {
...
+            if let Some(end_key) = end_key {
+                self.key_range.start..end_key
+            } else {
+                self.key_range.clone()
+            },
             self.lsn,
             metadata.len(),
         );
 
+        // TODO: when end_key is specified, verify that the end key is in range
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/image_layer.rs:917]: can you implement this TODO? Maybe via an assert or error log or something?

File: pageserver/src/tenant/storage_layer/split_writer.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    pub async fn put_image(
+        &mut self,
+        key: Key,
+        img: Bytes,
+        tline: &Arc<Timeline>,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<()> {
+        let addition_size_estimation = KEY_SIZE as u64 + img.len() as u64;
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/split_writer.rs:61]: Maybe add a comment that this is an upper bound due to compression, but it's okay to cut a little bit smaller layer files than needed.

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 4]
```diff
@@ -988,7 +1012,22 @@ impl ImageLayerWriter {
         timeline: &Arc<Timeline>,
         ctx: &RequestContext,
     ) -> anyhow::Result<super::ResidentLayer> {
-        self.inner.take().unwrap().finish(timeline, ctx).await
+        self.inner.take().unwrap().finish(timeline, ctx, None).await
+    }
+
+    #[cfg(test)]
+    /// Finish writing the image layer with an end key, used in [`super::split_writer::SplitImageLayerWriter`]
```
Comment (arpad-m) [inline pageserver/src/tenant/storage_layer/image_layer.rs:1019]:
```suggestion
    /// Finish writing the image layer with an end key, used in [`super::split_writer::SplitImageLayerWriter`]. The end key determines the end of the image layer's covered range and is exclusive.
```

**Review by arpad-m — APPROVED** [review]

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/tenant/storage_layer/image_layer.rs:917]: This requires we copy the last-written key somewhere when `put_image` is called, and then verify it at the end. I plan to add that in a separate pull request with `#[cfg(test)]`.

Comment (skyzh) [reply pageserver/src/tenant/storage_layer/image_layer.rs]: added in https://github.com/neondatabase/neon/pull/8574/commits/bb60bdfa68fa6028518607940b5a52e63a0e9b29, turned out to be a small change

**Pull request merged**